### ISSUE TITLE
Feature request: output file to replace existing

### ISSUE THREAD

**Issue by alexandercerutti** [axelpale/genversion#22]
Hi, thanks for this tool! I've been using it for over three years!

I know it's been a while since the last release, but I'd like to ask you if is there any chance if you could add an option to force the creation of the target file.

The idea is to allow importing the dummy file that contains a placeholder while developing and the correct one once a library is released. Sometimes we link the library to a webpack based application, without passing through through the tipical release pipeline (so, webpack handles its building).

Currently, by doing this, it outputs:

```js
File lib/esm/version.js is not generated by genversion and therefore will not be replaced. Please ensure that the file can be destroyed and remove it manually before retry.
```

Assuming that `version.js` is `<target>` and the building output is `lib/`:

```js
// index.js
export { version } from "./version";
```

```js
// version.js
export const version = "$dev";
```

and then, after running `genversion` (`$ pnpm genversion -esd <target>`), the file should get replaced correctly.

I know it could be possible to run `rm -rf <target>`, but we should have the same path everywhere and that would be prone to mistakes in the future.

I have something like this command:

```json
{
	"scripts": {
		"build:cjs": "pnpm tsc -p tsconfig.build.cjs.json && pnpm genv -- lib/cjs/version.js",
		"build:esm": "pnpm tsc -p tsconfig.build.esm.json && pnpm genv -- lib/esm/version.js",
		"genv": "pnpm genversion -esd"
	}
}
```

Having the `rm -rf` would make us to create two different commands, each to run `genversion` on a path.

I'm also available to work on it and open a PR.

Let me know. Thank you!

**Comment by axelpale** [comment]
Hi! Thanks for the feedback and suggestion. If I understood correctly, you would like to have a placeholder version.js that the genversion is able to overwrite when needed.

Currently genversion overwrites the file only if the first line of the file matches the regexp pattern `/generated by genversion/i`. The original purpose of this safeguard was to avoid accidents where a misconfiguration might cause genversion to overwrite an important file in the project or elsewhere. Usually projects are under version control and therefore such accident would not be a big issue because likely easily detected. But still, the possibility for damage would exist if there were no safeguard.

You possibly picked the hint already; an easy way to solve this is to modify the first line of your placeholder file to look like it was created by genversion. In other words, just replace `// version.js` with `// Generated by genversion.` so that genversion knows the file is safe to overwrite.

**Comment by axelpale** [comment]
Happy new year! Finally I got some free time to work on genversion.

I am planning for two features related to the issue. The first is the introduction of `-f --force` flag that overwrites the target file regardless the contents of the first line or the file in general. The second is an ability to specify a custom template with `-t --template` flag (#23).

I assume that the first, `--force`, can solve your initial issue of genversion refusing to overwrite the file regardless of the extra code added in typescript compilation. However, the overwriting removes those extra lines which can be important to you or the compilation process. Although I would instruct genversion to be run before any compilation takes place, I can understand that not all pipelines can allow that. For that, the second feature, `--template`, can bring some help by letting you to customise the file generated with genversion. With a custom template you could include and match the lines produced by the typescript compilation. There is the long-term maintenance downside such that you might need to update the custom template if your complication process changes for example due to a new typescript version. Still, the template feature at least provides _a_ solution.

I am planning to release the both features in the next genversion release v3.2, hopefully within a week or so. If you have any comments on these features, regarding your use case or in general, feel free to share.

**Comment by alexandercerutti** [comment]
Hi @axelpale, happy new year to you!
Thanks for your reply and the work on this library!

The main issue that comes to my mind about running `genversion` before anything else, is that it changes a _working directory_ file. Which translates, for all the cases out there that (just like me) should run `genversion` before any compilation step, into the need of adding a script that resets the git head before publishing the package (as it is required).

That's not very convenient and might bring to a few downsides and issues. Publishing a package with a dirty git working directory can also be skipped, but that might bring some unwanted and not ignored files (perhaps temporary but required) into the package itself. Quite a big deal IMHO.

So, the best action here is to run `genversion` on a pre-built file that is `gitignore`d, or in a folder just made for distribution (which usually already have some commands for cleaning up or ignoring the folder).

___

Then, moving to the proposal of `--force`, it might have less and less sense (exclusively for my use case) given what happens with Typescript compilation (which, still, I'm not very sure the code it outputs is needed in all the cases).

I find the proposal in #23 very very interesting, but I still have some doubts on how this could cover my use case.

Given the Typescript-outputted code, it might have much sense to create something like a "block" that `genversion` recognizes and goes to act on, even after third parties compilation.

Hence, from:

```ts
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.version = void 0;
// Generated by genversion.
exports.version = "x.y.z";
//# sourceMappingURL=version.js.map
```

To:

```ts
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.version = void 0;
// genversion:start
exports.version = "x.y.z";
// genversion:end
//# sourceMappingURL=version.js.map
```

**But I think we might not be very sure other compilers won't change the code inside the "block".** One example that comes to my mind (ofc, to be tested and evaluated) is the case of a project that builds with Webpack (or any other bundler), `ts-loader` and `TerserPlugin` that removes the comments (therefore the block would get destroyed. To avoid this, `genversion` should be run in webpack compilation before TerserPlugin, which is not an easy task by the way).

I'm maybe of the idea that using an AST Parser to parse the file and lookup for the variable "version", change its value and build the file back could be the best way to avoid most of the problems. You could just read the variable name as an input!

If you could have an AST Parser that does this job for you, you wouldn't have to care about:
 - Code output (cjs or esm, except of how you should look for the property in the AST)
 - Code format (`prettier` or `eslint`, double column, single column, single quotes, double quotes: all of them could and would be handled externally before or after `genversion` execution).

Having an AST Parser would therefore make the whole `template` feature (which feels to be a bit an overkill and a bit much complex to me) **useless**. And you still might want to keep the `templateEngine` concept to let people have a different AST parser, perhaps one for EJS (you should therefore expose a known interface through which you do expect specific things).

But I think that having the possibility to specify a `templateEngine` should be something that should be evaluated later and **only if needed or requested**, because I feel it could take a lot of effort to create and I don't know if there are other people out there that use a different format.

In general, my idea is that `genversion` should not care about the user / project settings, as they feel to be out-of-scope matters.

By having an AST Parser, you could deprecate flags such as `--semi` or `--double` and `--strict` and you would not have the issues that you described in #23 to make templates fit with the current existing flags. Deprecating those flags, would make your way ready to introduce a possible `templateEngine`.

I don't know if you agree with my idea and what's yours. I'm genuinely curious.

Again, thank you!

**Comment by axelpale** [comment]
Thanks @alexandercerutti for your thoughtful and honest view on this. It definitely makes sense for genversion to replace a variable in a target file – under version control or not – instead of creating or changing a file that is likely under version control. I will take this point and the others into consideration when planning for the upcoming release. The work continues...

**Comment by alexandercerutti** [comment]
@axelpale I had a few spare time today and I attempted to implement a little POC with an AST parser with string replacement and file re-build, for both CJS and ESM, if you are interested.

Of course, some corner case might not be covered, but this is a good start to me.

https://codesandbox.io/p/devbox/l76k99

**Comment by axelpale** [comment]
Hi @alexandercerutti and congrats for the good work. The AST parser approach is interesting and definitely would suit many projects very well. The AST parser solves the formatting in very different but powerful way. I am hesitant, however, for its complexity when compared to good ol' string replace. Maybe your AST parser application to meta data insertion is worth its own package because it gives lots of power and possibilities for its complexity. I can easily imagine a tool, say "plaster" (note *ast* :D), that allows devs to insert properties to source bundles with simple assignments like `plaster .version=$version dist/bundle.js`.

While AST parser is great in ensuring correct structure, style, and format, I assume it requires that the file to parse is perfectly formatted. Parsing takes time and with large bundles things might slow down. After all, genversion is just a tiny piece in long pipelines, it should be very simple under the hood too and without much requirements placed for the host project.

My current working plan for genversion is to implement `--variable <string>` parameter and `--substitute` flag. The given `<string>` is rendered in the generated module file instead of the actual version tag. When the substitute flag is present, that string is being replaced with the version tag. The substitute flag kind of repurposes genversion's module generation as template generation. The user who wants to insert the actual version into post-build bundles needs to run genversion only once to generate the template-ish module, let that module be compiled into the bundle, and then add a `genversion --substitute` call somewhere along the pipeline. No changes into version control after the initial generation, and best of all users can easily modify the generated template to their needs.

**Comment by alexandercerutti** [comment]
> The AST parser solves the formatting in very different but powerful way

_Au contraire_, actually (😄). I thought and hoped an AST could have given access to the single tokens used to format (semi-colons, commas...), but I was wrong.

> While AST parser is great in ensuring correct structure, style, and format, I assume it requires that the file to parse is perfectly formatted.

Does it? AST seems to not care about formatting until it is a valid JS (if said so). So people are free to format the file as they wish, with the tool they wish.

What I see, though, is that `escodegen` owns some formatting settings and some defaults that might not match with the original settings, so additional checks (probably based on some heuristics?) could be performed to verify which settings should be set when outputting. Otherwise, auto-detection based on the most common formatting tools' settings could be set (e.g. reading `.prettierrc` or `.eslintrc`, or something like that and merging them). 

I agree with you on the parsing time, but I think that if a package is provided with specific instructions (e.g. "creating a file named `version.js` containing an exported variable named `VERSION`"), most people will follow the suggested approach and fewer will create something complex. So, this can cut out most of the corner cases' discovery and complexity in most of the usages. Might be wrong though.

> My current working plan for genversion is to implement --variable <string> parameter and --substitute flag. The given <string> is rendered in the generated module file instead of the actual version tag. When the substitute flag is present, that string is being replaced with the version tag. The substitute flag kind of repurposes genversion's module generation as template generation. The user who wants to insert the actual version into post-build bundles needs to run genversion only once to generate the template-ish module, let that module be compiled into the bundle, and then add a genversion --substitute call somewhere along the pipeline. No changes into version control after the initial generation, and best of all users can easily modify the generated template to their needs.

I've read #26, and I think this is fine. Also, these options could interop pretty much correctly with the AST (I mean, I don't think they could break anything, just in case).

What's not very clear to me is why one "`needs to run genversion only once to generate the template-ish module`". I mean, couldn't we users just create our file and, once given the right settings, `genversion` just replaces them? Also about what you wrote later: "and best of all users can easily modify the generated template to their needs.". I'm thinking about not creating a tailored file for the version but hardcoding the variable into the index barrel file that exports everything.

Thank you!

**Comment by axelpale** [comment]
Thanks again for the sharp and thoughtful input.

> AST seems to not care about formatting until it is a valid JS (if said so)

Sorry, I misused formatting as a umbrella term for all structure, including valid JS. Although valid JS likely covers 98% of the cases, it is still a strong assumption and a requirement we are better without if possible. _I know that I do not know_ where people are using genversion and I can begin to imagine some exotic pipeline steps that break JS validity temporarily. Also, users of JS-flavoured languages that are not valid JS (coffeescript) can experience issues and would need to be unnecessary careful when placing genversion in their pipeline.

> "needs to run genversion only once to generate the template-ish module". I mean, couldn't we users just create our file

You are correct, the generation step would not be necessary for users who would like to create their own file. Also, I am glad that you pointed out the possibility for users to just hardcoding the variable into the root index file or wherever needed. Personally, I prefer separation of concerns a dedicated file brings, but I can see the value in not forcing users to do that. I need to consider the backward-compatibility in genversion usage, thus we cannot just drop the generation feature and expect everybody to migrate to variables overnight. Therefore I need to find a stable project state that is backward-compatible but still allows the variable approach. However, if the variable approach takes off well, it might take over the generation approach as a primary way to use genversion at some point in the future.

**Comment by alexandercerutti** [comment]
> Also, users of JS-flavoured languages that are not valid JS (coffeescript) can experience issues and would need to be unnecessary careful when placing genversion in their pipeline.

I don't know if others feel the same, but I feel the most intuitive usage of `genversion` is after every library compilation step.

But I think the doubt could be "easily" resolved by "creating an opinionated manifesto", a set of boundaries and guidelines that you'd like people to adhere to and that, if surpassed, could lift you up from the responsibility of providing any change to the program or even support, if a reported issue does not provide valid points to change the manifesto and the code.

One of the points of such manifesto, could be that using genversion after any compilation step is highly recommeded. Of course, such manifesto would contain rules and indications you feel to be right.

Once a manifesto is provided, entropy tends to get pretty much reduced IMHO. Your current manifesto is what you wrote in the paragraph "Try it out".

> I can begin to imagine some exotic pipeline steps that break JS validity temporarily

No please, don't do that lol. Unless you want to end in the Alice in Wonderland's rabbit hole 😄

> Personally, I prefer separation of concerns a dedicated file brings, but I can see the value in not forcing users to do that.

Definitely agree, but the fact of putting the variable inside the index file is still a possibility.

> I need to consider the backward-compatibility in genversion usage, thus we cannot just drop the generation feature and expect everybody to migrate to variables overnight.

Of course. The most critical thing is the usage through `npx` without prior installation. You might just want to be sure that eventual deprecated and removed options do not break the flows, if you decide to bump the major semver.

> Therefore I need to find a stable project state that is backward-compatible but still allows the variable approach.

I'd keep `"VERSION"` as default lookup string, allow the new optional flag `--variable` and always allow `--substitute / -s` (or maybe `--replace / -r` ?).

I honestly don't see any difference between "generation" approach and "variable" approach.

I might be wrong, but if you see the flow as _"reading file" -> "changing" -> "emitting"_, you might just want to change _"reading file"_ to _"reading file or fallback"_, and the fallback would just use the name in `--variable` or a default one `"VERSION"` (to be retro-compatible).

After all, users that are already generating the file on demand, they are already fine with what is outputted right now, so having a retro-compatible fallback is important.

**Comment by axelpale** [comment]
You really help me think this through, thank you for that!

> I feel the most intuitive usage of genversion is after every library compilation step

Of course a major audience for genversion are those who compile and bundle their client-side code. However, there exists many projects, especially libraries, that do well without. For example, genversion itself is an example of a project that uses genversion but does not have any compilation steps. Other examples of my own are [affineplane](https://github.com/axelpale/affineplane) and [tapspace](https://github.com/taataa/tapspace). The latter has a build pipeline for standalone bundles but is also useable as-is, and therefore Tapspace's raw codebase cannot contain any unfinished business such as variables to be replaced. Furthermore, with the advent of native browser support for ECMAScript modules, there must be increasing number of projects that do not trouble themselves with a build pipeline. Therefore the ability to generate a finished version module without any variable-trickery is important.

> I'd keep "VERSION" as default lookup string, allow the new optional flag --variable and always allow --substitute / -s (or maybe --replace / -r ?).

I am thinking of using "GENVERSION_VERSION" as the default variable name if `--variable` is present and without a value. It is way too likely that a large bundle accidentally contains "VERSION" somewhere. Regarding the flag name, I did consider `--replace` but I feel it to be too easily misunderstood as file replacement instead of variable substitution in the context of genversion. I also considered `--assign` as in _variable assignment_ but its meaning would not be accurate because we are not changing the value of the variable but substituting the variable with a literal. I wish there would be a simpler word than "substitute" that is still accurate and not easily confused with file operations.

**Comment by alexandercerutti** [comment]
> However, there exists many projects, especially libraries, that do well without.

Well, yeah, in the distribution pipeline, the compilation is purely optional.

> I am thinking of using "GENVERSION_VERSION" as the default variable name if --variable is present and without a value. 

Wouldn't be more explicit and understandable to throw an error if `--variable` is set as empty? It might mean that the user forgot to set the name. Otherwise one might go to check why things are not working and end up then looking inside the built file only then to discover something that, perhaps, it didn't read on the documentation (their fault, but still time taking to realize). I always prefer explicit things over assumptions.

> It is way too likely that a large bundle accidentally contains "VERSION" somewhere

Usually, bundles like Webpack split the modules into functions. The only corner case that comes to my mind could be UMD, but I'm not even sure this is a real problem right now.

> Regarding the flag name, I did consider --replace but I feel it to be too easily misunderstood as file replacement instead of variable substitution in the context of genversion. I also considered --assign as in variable assignment but its meaning would not be accurate because we are not changing the value of the variable but substituting the variable with a literal. I wish there would be a simpler word than "substitute" that is still accurate and not easily confused with file operations.

You have a valid point. I could also suggest _--apply_ (which is great, but might confuse those not using it to think they are not actually applying the change) or _--supersede_ (which is similar to _--substitute_ but takes the place of) or a more verbose `--change-existing`.

**Comment by axelpale** [comment]
> Wouldn't be more explicit and understandable to throw an error if --variable is set as empty?

You are right. I agree that explicit is better than implicit, thus there should not be a default variable name.

> I could also suggest --apply or --supersede

Good suggestions. The following might help. I came to understand better what you wrote previously, including:

> I honestly don't see any difference between "generation" approach and "variable" approach.

I do agree now that generating a version module that includes a placeholder variable does not serve much purpose. Users that just want to substitute variables after build can easily hardcode the variable wherever they like and in whatever style they want. Therefore, as I already wrote in #26, I am dropping `--substitute` idea altogether and just make `--variable` to indicate substitution in the target file. Here is an example that would just replace the variable in the target bundle file:

```
$ genversion --variable PACKAGEVERSION dist/bundle.js
```

Taking into consideration that --variable must have a value, I again feel tempted to rename the flag as `--replace`. The result almost reads like "replace this in that":

```
$ genversion --replace PACKAGEVERSION dist/bundle.js
```

To give an example with multiple properties and your `--apply` suggestion:

```
$ genversion --property name,version --apply PKGNAME,PKGVERSION dist/bundle.js
```

**Comment by alexandercerutti** [comment]
> You are right. I agree that explicit is better than implicit, thus there should not be a default variable name.

I mean, you might choose to have a default only if `--variable` is not specified. By doing this way, you can assure retro-compatibility.

___

I think I like this.
A possible improvement could be that `--property` and `--apply` merge into a (I think) docker-like mapping.

```sh
$ genversion --apply name:PKGNAME,version:PKGVERSION dist/bundle.js
```

By doing this, I think you could remove any ambiguity between the order of the two parameters.

I have a doubt now. By having the concept of remapping like `property`, isn't this going further than the simple `genversion`?

Also, what I don't have clear is if `--apply` is now meant to replace `--variable` for multiple params or add parameters.

You could have `--apply` or just `--variable` with the same behavior described above, which behaves in these ways:

- Looks for a default "VERSION" and replaces with package.json's 'version', which is more than enough for most of the cases out there, IMHO:

```sh
$ genversion dist/bundle.js
```

- Goes on Error for missing parameter for `--apply`:

```sh
$ genversion --apply dist/bundle.js
```

- Gets 'version' from 'package.json' and seeks for PKGVERSION in the file:

```sh
$ genversion --apply version:PKGVERSION dist/bundle.js
```

- Gets all the specified fields from 'package.json' and remaps them

```sh
$ genversion --apply name:PKGNAME,version:PKGVERSION dist/bundle.js
```

An optional thing that might interest you, could be the possibility of specifying such parameters in a JSON map in `package.json` itself (you are already reading that so, why not?). You could also take advantage of this package if you want to exaggerate 😄 https://github.com/cosmiconfig/cosmiconfig

___

> In any case, you have been a tremendous help :)

🚀 🚀 🚀

**Comment by axelpale** [comment]
> A possible improvement could be that --property and --apply merge into a (I think) docker-like mapping

The docker-like mapping is a sound improvement. Still, `--property` needs to be available for those who do not `--apply` to variables but generate the module. I will attempt to make `--apply` work both ways, with `--property` or the docker-like mapping, as in the snippets further below.

> I have a doubt now. By having the concept of remapping like property, isn't this going further than the simple genversion?

Yes, the customisation of properties expands possible use cases of genversion outside of mere package version. However, I view this feature more as a supplement for version getting than complete repurposing of the tool. According to #21 there are cases where the package name is a desirable supplementary datum, for example to be displayed next to the version in log output. If the feature makes lives of genversion users easier, I would go for it.

Still, I am a bit worried the new possibility to pick properties can turn out to be a can of worms for maintaining genversion. I can already see problems with applying non-string properties from package.json, like "keywords". While it sounds like a simple matter of replacing a placeholder variable, say `PKGKEYWORDS`, with the output of `JSON.stringify(pjson.keywords)`, in reality it is shaky to hardcode such unquoted variable PKGKEYWORDS in the source code without linters or compilation going red, am I right?

Application of string properties is much easier, because the hardcoded `export const version = "PKGVERSION"` would just turn to `export const version = "1.2.3"` upon replacement. Compare this with `export const keywords = "PKGKEYWORDS"` that would become the troublesome `export const keywords = "["hello", "world"]"`. If we remove hardcoded double quotes like `export const keywords = PKGKEYWORDS`, then the PKGKEYWORDS becomes a undefined JS variable from linters' perspective, requiring users to configure the variable as a global variable or to be ignored by linter/compiler. Still even more wormy, the result would become `export const keywords = ["hello", "world"]` which is valid but causes users to begin to request formatting of the JSON.stringify output, like usage of single quotes instead of doubles. While possible, it's tedious from my maintainer point of view.

For these reasons, I need to limit the set of applicable properties to those that have string value. In effect, this limitation limits possible new use cases, emphasising the mere supplementary role of `--property` with respect to version-getting.

> $ genversion dist/bundle.js

Regarding the default behavior, that I wont be changing from module generation. The command above will generate the most basic version module file at dist/bundle.js instead of variable replacement. Changing that promise would constitute to a new major version and would probably still annoy current users.

However, I will keep the `version` property as the default. Therefore the most minimal usage of `--apply` would become:

$ genversion --apply PKGVERSION dist/bundle.js

which would be equivalent to:

$ genversion --apply version:PKGVERSION dist/bundle.js

as well as:

$ genversion --property version --apply PKGVERSION dist/bundle.js

> the possibility of specifying such parameters in a JSON map in package.json itself

That is a good suggestion for the future. However, designing that feature would be much easier if genversion were only about replacing variables and not about generation of modules. It would be a great feature of that hypothetical spin-off tool *plaster*.

**Comment by alexandercerutti** [comment]
> Still, --property needs to be available for those who do not --apply to variables but generate the module. I will attempt to make --apply work both ways, with --property or the docker-like mapping, as in the snippets further below.

So `--property` will generate the file to include the specified fields? Did I understand correctly? I'd keep only `--apply`. Having two parameters that do almost the same things could be source of confusion.

> Still, I am a bit worried the new possibility to pick properties can turn out to be a can of worms for maintaining genversion. I can already see problems with applying non-string properties from package.json, like "keywords". While it sounds like a simple matter of replacing a placeholder variable, say PKGKEYWORDS, with the output of JSON.stringify(pjson.keywords), in reality it is shaky to hardcode such unquoted variable PKGKEYWORDS in the source code without linters or compilation going red, am I right?

I think that people could expect the value that can be found in the JSON. Say I (as a user) select `keywords`, I might expect to have the array replaced in the code, instead of the string. It might be a bit counter-intuitive to have the string instead. The same is true with any other kind of object. You should check the type of parameter inside the `package.json` and choose what action to take.

> While possible, it's tedious from my maintainer point of view.

I agree with all you have written up to this point. This could be tedious, error-prone, and confusing. You might want either to restrict replacement only to strings, as you said or do that just for the version.

**Comment by alexandercerutti** [comment]
> I am thinking --property being for generation, only to choose the set of properties to bake in the generated file, while --apply is for the in-place replacement so that its value maps a set of properties to the variables to be replaced. There would be no --substitute flag or similar to activate replacement, but --apply alone would make genversion to replace instead of generate.

What's the difference between generating and applying? I mean, generation could be considered a fallback for failing to read a file, isn't it?

> However, I realised there is a big problem for --apply in general what makes its use sketchy at best, namely the [source maps](https://www.mattzeunert.com/2016/02/14/how-do-source-maps-work.html). If we modify a bundle in any way that changes the code length even by a single character, that would break mappings in source maps. I hope that every project that compiles and bundles the code also uses some form of source maps (personally I do) at least during development. Letting a simple tool like genversion to mess that up is a big no-no. And I can guarantee you that I, as a genversion developer, wont begin to learn how to modify source maps :D

Yeah, you are right. Source maps would break. Said I think that a solution should be as easy as possible, and also said that I think bundlers should not be used for libraries, I think one of the rules of your manifesto should be that `genversion` should be applied only on libraries and (only if needed) before any kind of bundling, not for final projects.

Changing a value inside a bundle makes almost no sense IMHO: `DefinePlugin` for webpack as well as other integrated tools for the other bundlers, already exist and they do exactly what you are describing, without compromising the source maps.

Learning source maps? No thx! :D

**Comment by axelpale** [comment]
Hi @alexandercerutti ! Just a moment ago I released genversion v3.2.0 with many features which you initiated or helped me to think through. I am very grateful for that. Unfortunately, I could not wrap my head around the `--apply` nor `--variable` flags, and therefore I had to discard them for the time being, at least until I can see their purpose more clearly.

That said, I suppose your original issue can be solved with the new flags `--force` and `--template`. Thus, I will close this long issue now and celebrate the new release. Still, feel free to open new issues if those new flags do not serve your use case well, or if you do have more good ideas up in your sleeve!

It can take a while before I give genversion another development sprint, except for minor bugfixes. Still, the work will continue at some point. Thank you for all the help! So long, take care! 🎸
